### DISCUSSION TITLE
Use R2R for RAG backend

### DISCUSSION THREAD

**Discussion by emrgnt-cmplxty** [open-webui/open-webui#1333] [General]
**Problem:**
Open WebUI's current RAG integration is in the alpha phase and may face occasional issues.

**Solution:**
Integrate [R2R](https://github.com/SciPhi-AI/R2R), a production-ready RAG framework, as the backend for Open WebUI's RAG feature. This will improve reliability, performance, extensibility, and maintainability.

**Steps:**
1. Install R2R and its dependencies in Open WebUI.
2. Configure R2R's environment variables.
3. Modify Open WebUI's RAG implementation to use R2R's pipelines.
4. Manage data flow between Open WebUI and R2R.
5. Test the integration thoroughly.

**Benefits:**
- Leverage R2R's flexibility to integrate with various vector databases, LLMs, and embedding models.
- Utilize R2R pipeline templates to allow end users to quickly deploy bespoke RAG pipelines.
- Add support for logging / tracing out of the box.

By integrating R2R with Open WebUI, users can enjoy a more reliable and performant RAG experience, while developers can focus on further enhancements and customizations.

**Comment by Killswitchz** [comment]
Can you please share a bit more details how you can do the integration?

**Comment by thecosmicskye** [comment]
Updates on this?

**Comment by justinh-rahb** [comment]
Seeing as how we already have a RAG pipeline that mostly works, I'm gonna be honest and say this probably won't receive much prioritization by the core devs. Welcome to contributions though.

**Reply by mMabeck** [reply]
But isn't the current implementation very limited?
An option to just search through all the documents/knowledge base "behind-the-scenes" and retrieve different chunks would be very useful.

**Reply by justinh-rahb** [reply]
Since last week, we've now introduced [Pipelines](https://github.com/open-webui/pipelines), which is specifically meant for implementing custom pipelines.
